man/zabbix_js.man
.\" Start URL macro. Copied from an-ext.tmac for portability
.de UR
.  ds m1 \\$1\"
.  nh
.  if \\n(mH \{\
.    \" Start diversion in a new environment.
.    do ev URL-div
.    do di URL-div
.  \}
..
.\" End URL macro. Copied from an-ext.tmac for portability
.de UE
.  ie \\n(mH \{\
.    br
.    di
.    ev
.
.    \" Has there been one or more input lines for the link text?
.    ie \\n(dn \{\
.      do HTML-NS "<a href=""\\*(m1"">"
.      \" Yes, strip off final newline of diversion and emit it.
.      do chop URL-div
.      do URL-div
\c
.      do HTML-NS </a>
.    \}
.    el \
.      do HTML-NS "<a href=""\\*(m1"">\\*(m1</a>"
\&\\$*\"
.  \}
.  el \
\\*(la\\*(m1\\*(ra\\$*\"
.
.  hy \\n(HY
..
.TH ZABBIX_JS 1 "2022\-01\-06" Zabbix
.SH NAME
zabbix_js \- Zabbix JS utility
.SH SYNOPSIS
.B zabbix_js \-s
.I script\-file
.B \-p
.I input\-param
.RB [ \-l
.IR log\-level ]
.RB [ \-t
.IR timeout ]
.br
.B zabbix_js \-s
.I script\-file
.B \-i
.I input\-file
.RB [ \-l
.IR log\-level ]
.RB [ \-t
.IR timeout ]
.br
.B zabbix_js \-h
.br
.B zabbix_js \-V
.SH DESCRIPTION
.B zabbix_js
is a command line utility that can be used for embedded script testing.
.SH OPTIONS
.IP "\fB\-s\fR, \fB\-\-script\fR \fIscript\-file\fR"
Specify the file name of the script to execute. If '-' is specified as file name, the script will be read from stdin.
.IP "\fB\-p\fR, \fB\-\-param\fR \fIinput\-param\fR"
Specify the input parameter passed to the script. Parameter is passed verbatim as a string and is available in the script via the 'value' variable. Mutually exclusive with -i.
.IP "\fB\-i\fR, \fB\-\-input\fR \fIinput\-file\fR"
Specify a file containing the input parameter passed to the script. The file contents are passed verbatim as a string and are available in the script via the 'value' variable. If '-' is specified as file name, the input will be read from stdin. Mutually exclusive with -p.
.IP "\fB\-w\fR, \fB\-\-webdriver\fR \fIurl\fR"
Specify the webdriver URL.
.IP "\fB\-l\fR, \fB\-\-loglevel\fR \fIlog\-level\fR"
Specify log level (Range: 0-5).
.IP "\fB\-t\fR, \fB\-\-timeout\fR \fItimeout\fR"
Specify the timeout in seconds. Valid range: 1\-600 seconds (default: 10)
.IP "\fB\-h\fR, \fB\-\-help\fR"
Display this help and exit.
.IP "\fB\-V\fR, \fB\-\-version\fR"
Output version information and exit.
.SH "EXAMPLES"
\fBzabbix_js \-s script-file.js \-p example \fR
.SH "SEE ALSO"
.UR https://www.zabbix.com/manuals
Documentation
.UE
.PP
.BR zabbix_agentd (8),
.BR zabbix_proxy (8),
.BR zabbix_sender (1),
.BR zabbix_server (8),
.BR zabbix_get (1),
.BR zabbix_agent2 (8),
.BR zabbix_web_service (8)


NEWS
Please see https://www.zabbix.com/documentation.php.


src/zabbix_js/zabbix_js.c
/*
** Copyright (C) 2001-2026 Zabbix SIA
**
** This program is free software: you can redistribute it and/or modify it under the terms of
** the GNU Affero General Public License as published by the Free Software Foundation, version 3.
**
** This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
** without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
** See the GNU Affero General Public License for more details.
**
** You should have received a copy of the GNU Affero General Public License along with this program.
** If not, see <https://www.gnu.org/licenses/>.
**/

#include "zbxlog.h"
#include "zbxgetopt.h"
#include "zbxembed.h"
#include "zbxmutexs.h"
#include "zbxstr.h"
#include "zbxnix.h"
#include "zbxnum.h"
#include "zbxbincommon.h"

ZBX_GET_CONFIG_VAR2(const char *, const char *, zbx_progname, NULL)
static const char	title_message[] = "zabbix_js";
static const char	syslog_app_name[] = "zabbix_js";
static const char	*usage_message[] = {
	"-s script-file", "-p input-param", "[-l log-level]", "[-t timeout]", NULL,
	"-s script-file", "-i input-file", "[-l log-level]", "[-t timeout]", NULL,
	"-h", NULL,
	"-V", NULL,
	NULL	/* end of text */
};

#define ZBX_SERVICE_NAME_LEN	64
char	zabbix_event_source[ZBX_SERVICE_NAME_LEN] = APPLICATION_NAME;
#undef ZBX_SERVICE_NAME_LEN

#define JS_TIMEOUT_MIN		1
#define JS_TIMEOUT_MAX		600
#define JS_TIMEOUT_DEF		ZBX_ES_TIMEOUT
#define JS_TIMEOUT_MIN_STR	ZBX_STR(JS_TIMEOUT_MIN)
#define JS_TIMEOUT_MAX_STR	ZBX_STR(JS_TIMEOUT_MAX)
#define JS_TIMEOUT_DEF_STR	ZBX_STR(JS_TIMEOUT_DEF)

static const char	*help_message[] = {
	"Execute script using Zabbix embedded scripting engine.",
	"",
	"General options:",
	"  -s,--script script-file      Specify the filename of script to execute. Specify '-' for",
	"                               standard input.",
	"  -i,--input input-file        Specify a file containing the input parameter passed to the",
	"                               script. The file contents are passed verbatim as a string",
	"                               and are available in the script via the 'value' variable.",
	"                               Specify '-' to read from standard input. Mutually",
	"                               exclusive with -p.",
	"  -p,--param input-param       Specify the input parameter passed to the script. Parameter",
	"                               is passed verbatim as a string and is available in the",
	"                               script via the 'value' variable. Mutually exclusive with -i.",
	"  -w,--webdriver url           Specify webdriver URL",
	"  -l,--loglevel log-level      Specify log level (Range: 0-5).",
	"  -t --timeout timeout         Specify the timeout in seconds. Valid range: " JS_TIMEOUT_MIN_STR "-"
			JS_TIMEOUT_MAX_STR " seconds",
	"                               (default: " JS_TIMEOUT_DEF_STR " seconds)",
	"  -h --help                    Display this help message",
	"  -V --version                 Display version number",
	"",
	"Example:",
	"  zabbix_js -s script-file.js -p example",
	NULL	/* end of text */
};

/* long options */
struct zbx_option	longopts[] =
{
	{"script",			1,	NULL,	's'},
	{"input",			1,	NULL,	'i'},
	{"param",			1,	NULL,	'p'},
	{"webdriver",			1,	NULL,	'w'},
	{"loglevel",			1,	NULL,	'l'},
	{"timeout",			1,	NULL,	't'},
	{"help",			0,	NULL,	'h'},
	{"version",			0,	NULL,	'V'},
	{0}
};

/* short options */
static char	shortopts[] = "s:i:p:hVl:t:w:";

/* end of COMMAND LINE OPTIONS */

static char	*read_file(const char *filename, char **error)
{
	char	buffer[4096];
	int	fd;
	ssize_t	n;
	char	*data = NULL;
	size_t	data_alloc = 0, data_offset = 0;

	if (0 != strcmp(filename, "-"))
	{
		if (-1 == (fd = open(filename, O_RDONLY)))
		{
			*error = zbx_strdup(NULL, zbx_strerror(errno));
			return NULL;
		}
	}
	else
		fd = STDIN_FILENO;

	while (0 != (n = read(fd, buffer, sizeof(buffer))))
	{
		if (-1 == n)
		{
			if (fd != STDIN_FILENO)
				close(fd);
			zbx_free(data);
			*error = zbx_strdup(NULL, zbx_strerror(errno));
			return NULL;
		}
		zbx_strncpy_alloc(&data, &data_alloc, &data_offset, buffer, (size_t)n);
	}

	if (fd != STDIN_FILENO)
		close(fd);

	return data;
}

/******************************************************************************
 *                                                                            *
 * Purpose: executes command (script in form of a text)                       *
 *                                                                            *
 * Parameters: command          - [IN] command in form of a text              *
 *             param            - [IN] script parameters                      *
 *             timeout          - [IN] timeout for the execution (seconds)    *
 *             config_source_ip - [IN]                                        *
 *             webdriver        - [IN] webdriver URL (optional)
 *             result           - [OUT] result of an execution                *
 *             error            - [OUT] error message                         *
 *             max_error_len    - [IN] maximum length of an error             *
 *                                                                            *
 * Return value: SUCCEED                                                      *
 *               FAIL                                                         *
 *                                                                            *
 ******************************************************************************/
static int	execute_script(const char *command, const char *param, int timeout, const char *config_source_ip,
		const char *webdriver, char **result, char *error, size_t max_error_len)
{
	int		size, ret = SUCCEED;
	char		*code = NULL, *errmsg = NULL;
	zbx_es_t	es;

	zabbix_log(LOG_LEVEL_DEBUG, "In %s()", __func__);

	zbx_es_init(&es);
	if (FAIL == zbx_es_init_env(&es, config_source_ip, &errmsg))
	{
		zbx_snprintf(error, max_error_len, "cannot initialize scripting environment: %s", errmsg);
		zbx_free(errmsg);
		ret = FAIL;

		goto failure;
	}

	if (NULL != webdriver && FAIL == zbx_es_init_browser_env(&es, webdriver, &errmsg))
	{
		zbx_snprintf(error, max_error_len, "cannot initialize Browser object: %s", errmsg);
		zbx_free(errmsg);
		ret = FAIL;

		goto failure;
	}

	if (SUCCEED != zbx_es_globals_make_readonly(&es, &errmsg))
	{
		zbx_snprintf(error, max_error_len, "cannot initialize read-only environment: %s", errmsg);
		zbx_free(errmsg);
		ret = FAIL;

		goto failure;
	}

	if (FAIL == zbx_es_compile(&es, command, &code, &size, &errmsg))
	{
		zbx_snprintf(error, max_error_len, "cannot compile script: %s", errmsg);
		zbx_free(errmsg);
		ret = FAIL;

		goto out;
	}

	if (0 != timeout)
		zbx_es_set_timeout(&es, timeout);

	if (FAIL == zbx_es_execute(&es, NULL, code, size, param, result, &errmsg))
	{
		zbx_snprintf(error, max_error_len, "cannot execute script: %s", errmsg);
		zbx_free(errmsg);
		ret = FAIL;

		goto out;
	}
out:
	if (FAIL == zbx_es_destroy_env(&es, &errmsg))
	{
		zabbix_log(LOG_LEVEL_WARNING, "cannot destroy embedded scripting engine environment: %s", errmsg);
		zbx_free(errmsg);
	}

	zbx_free(code);
	zbx_free(errmsg);
failure:
	zabbix_log(LOG_LEVEL_DEBUG, "End of %s():%s", __func__, zbx_result_string(ret));

	return ret;
}

int	main(int argc, char **argv)
{
	int			ret = FAIL, loglevel = LOG_LEVEL_WARNING, timeout = 0;
	char			*script_file = NULL, *input_file = NULL, *param = NULL, ch, *script = NULL,
				*error = NULL, *result = NULL, script_error[MAX_STRING_LEN], *webdriver = NULL;
	zbx_config_log_t	log_file_cfg = {NULL, NULL, ZBX_LOG_TYPE_UNDEFINED, 0};

	/* see description of 'optarg' in 'man 3 getopt' */
	char			*zbx_optarg = NULL;

	/* see description of 'optind' in 'man 3 getopt' */
	int			zbx_optind = 0;

	const char		*config_source_ip = NULL;

	zbx_progname = get_program_name(argv[0]);

	zbx_init_library_common(zabbix_log_impl, zbx_get_log_level_impl, get_zbx_progname, zbx_backtrace);
#ifndef _WINDOWS
	zbx_init_library_nix(get_zbx_progname, NULL);
#endif
	/* parse the command-line */
	while ((char)EOF != (ch = (char)zbx_getopt_long(argc, argv, shortopts, longopts, NULL, &zbx_optarg,
			&zbx_optind)))
	{
		switch (ch)
		{
			case 's':
				if (NULL == script_file)
					script_file = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'i':
				if (NULL == input_file)
					input_file = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'p':
				if (NULL == param)
					param = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'w':
				if (NULL == webdriver)
					webdriver = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'l':
				loglevel = atoi(zbx_optarg);
				break;
			case 't':
				if (FAIL == zbx_is_uint_n_range(zbx_optarg, ZBX_MAX_UINT64_LEN, &timeout,
						sizeof(timeout), JS_TIMEOUT_MIN, JS_TIMEOUT_MAX))
				{
					zbx_error("Invalid timeout, valid range [" JS_TIMEOUT_MIN_STR ":"
							JS_TIMEOUT_MAX_STR "] seconds");
					zbx_exit(EXIT_FAILURE);
				}

				break;
			case 'h':
				zbx_print_help(zbx_progname, help_message, usage_message, NULL);
				ret = SUCCEED;
				goto clean;
			case 'V':
				zbx_print_version(title_message);
				ret = SUCCEED;
				goto clean;
			default:
				zbx_print_usage(zbx_progname, usage_message);
				goto clean;
		}
	}

	if (SUCCEED != zbx_locks_create(&error))
	{
		zbx_error("cannot create locks: %s", error);
		goto clean;
	}

	if (SUCCEED != zbx_open_log(&log_file_cfg, loglevel, syslog_app_name, zabbix_event_source, &error))
	{
		zbx_error("cannot open log: %s", error);
		goto clean;
	}

	if (NULL == script_file || (NULL == input_file && NULL == param))
	{
		zbx_print_usage(zbx_progname, usage_message);
		goto close;
	}

	if (NULL != input_file && NULL != param)
	{
		zbx_error("input and script options are mutually exclusive");
		goto close;
	}

	if (0 == strcmp(script_file, "-") && NULL != input_file && 0 == strcmp(input_file, "-"))
	{
		zbx_error("cannot read script and input parameters from standard input at the same time");
		goto close;
	}

	if (NULL == (script = read_file(script_file, &error)))
	{
		if (NULL != error)
			zbx_error("cannot read script file: %s", error);
		else
			zbx_error("cannot use empty script file");

		goto close;
	}

	if (NULL != input_file)
	{
		if (NULL == (param = read_file(input_file, &error)))
		{
			if (NULL != error)
				zbx_error("cannot read input file: %s", error);
			else
				zbx_error("cannot use empty input file");

			goto close;
		}
	}

	if (FAIL == execute_script(script, param, timeout, config_source_ip, webdriver, &result, script_error,
			sizeof(script_error)))
	{
		zbx_error("error executing script:\n%s", script_error);
		goto close;
	}
	ret = SUCCEED;
	printf("\n%s\n", result);
close:
	zbx_close_log();
#ifndef _WINDOWS
	zbx_locks_destroy();
#endif
clean:
	zbx_free(result);
	zbx_free(error);
	zbx_free(script);
	zbx_free(script_file);
	zbx_free(input_file);
	zbx_free(param);
	zbx_free(webdriver);

	return SUCCEED == ret ? EXIT_SUCCESS : EXIT_FAILURE;
}
